pkg/bloomcompactor/v2spec.go

```diff
@@ -0,0 +1,142 @@
+	logger  log.Logger
+
+	readWriterFn func() (v1.BlockWriter, v1.BlockReader)
+
+	tokenizer *v1.BloomTokenizer
+}
+
+// SimpleBloomGenerator is a foundational implementation of BloomGenerator.
+// It mainly wires up a few different components to generate bloom filters for a set of blocks
+// and handles schema compatibility:
+// Blocks which are incompatible with the schema are skipped and will have their chunks reindexed
+func NewSimpleBloomGenerator(
+	opts v1.BlockOptions,
+	store v1.Iterator[*v1.Series],
+	blocks []*v1.Block,
+	readWriterFn func() (v1.BlockWriter, v1.BlockReader),
+	metrics *Metrics,
+	logger log.Logger,
+) *SimpleBloomGenerator {
+	return &SimpleBloomGenerator{
+		opts:         opts,
+		store:        store,
+		blocks:       blocks,
+		logger:       logger,
+		readWriterFn: readWriterFn,
+		metrics:      metrics,
+
+		tokenizer: v1.NewBloomTokenizer(opts.Schema.NGramLen(), opts.Schema.NGramSkip(), metrics.bloomMetrics),
+	}
+}
+
+func (s *SimpleBloomGenerator) populate(series *v1.Series, bloom *v1.Bloom) error {
+	// TODO(owen-d): impl after threading in store
+	var chunkItr v1.Iterator[[]chunk.Chunk] = v1.NewEmptyIter[[]chunk.Chunk](nil)
+
+	return s.tokenizer.PopulateSeriesWithBloom(
+		&v1.SeriesWithBloom{
+			Series: series,
+			Bloom:  bloom,
+		},
+		chunkItr,
+	)
+}
+
+func (s *SimpleBloomGenerator) Generate(_ context.Context) (skippedBlocks []*v1.Block, results v1.Iterator[*v1.Block], err error) {
+
+	blocksMatchingSchema := make([]v1.PeekingIterator[*v1.SeriesWithBloom], 0, len(s.blocks))
+	for _, block := range s.blocks {
+		// TODO(owen-d): implement block naming so we can log the affected block in all these calls
+		logger := log.With(s.logger, "block", fmt.Sprintf("%#v", block))
+		schema, err := block.Schema()
+		if err != nil {
+			level.Warn(logger).Log("msg", "failed to get schema for block", "err", err)
+			skippedBlocks = append(skippedBlocks, block)
+		}
+
+		if !s.opts.Schema.Compatible(schema) {
+			level.Warn(logger).Log("msg", "block schema incompatible with options", "generator_schema", fmt.Sprintf("%#v", s.opts.Schema), "block_schema", fmt.Sprintf("%#v", schema))
+			skippedBlocks = append(skippedBlocks, block)
+		}
+
+		level.Debug(logger).Log("msg", "adding compatible block to bloom generation inputs")
+		itr := v1.NewPeekingIter[*v1.SeriesWithBloom](v1.NewBlockQuerier(block))
+		blocksMatchingSchema = append(blocksMatchingSchema, itr)
+	}
+
+	level.Debug(s.logger).Log("msg", "generating bloom filters for blocks", "num_blocks", len(blocksMatchingSchema), "skipped_blocks", len(skippedBlocks), "schema", fmt.Sprintf("%#v", s.opts.Schema))
+
+	// TODO(owen-d): implement bounded block sizes
+
+	mergeBuilder := v1.NewMergeBuilder(blocksMatchingSchema, s.store, s.populate)
+	writer, reader := s.readWriterFn()
+	blockBuilder, err := v1.NewBlockBuilder(v1.NewBlockOptionsFromSchema(s.opts.Schema), writer)
+	if err != nil {
+		return skippedBlocks, nil, errors.Wrap(err, "failed to create bloom block builder")
```

Comment:
Do we want to return `skippedBlocks` in case of an error? Does the consumer rely on them?